dlib/CMakeLists.txt

```diff
@@ -1,1039 +1,421 @@
-         global_optimization/global_function_search.cpp
-         filtering/kalman_filter.cpp
-         svm/auto.cpp
-         )
-
-      if(UNIX)
-         set(CMAKE_THREAD_PREFER_PTHREAD ON)
-         find_package(Threads REQUIRED)
-         list (APPEND dlib_needed_private_libraries ${CMAKE_THREAD_LIBS_INIT})
-      endif()
-
-      # we want to link to the right stuff depending on our platform.  
-      if (WIN32 AND NOT CYGWIN) ###############################################################################
-         if (DLIB_NO_GUI_SUPPORT)
-            list (APPEND dlib_needed_private_libraries ws2_32 winmm)
-         else()
-            list (APPEND dlib_needed_private_libraries ws2_32 winmm comctl32 gdi32 imm32)
-         endif()
-      elseif(APPLE) ############################################################################
-         set(CMAKE_MACOSX_RPATH 1)
-         if (NOT DLIB_NO_GUI_SUPPORT)
-            find_package(X11 QUIET)
-            if (X11_FOUND)
-               # If both X11 and anaconda are installed, it's possible for the
-               # anaconda path to appear before /opt/X11, so we remove anaconda.
-               foreach (ITR ${X11_INCLUDE_DIR})
-                  if ("${ITR}" MATCHES "(.*)(Ana|ana|mini)conda(.*)")
-                     list (REMOVE_ITEM X11_INCLUDE_DIR ${ITR})
-                  endif ()
-               endforeach(ITR)
-               list (APPEND dlib_needed_public_includes ${X11_INCLUDE_DIR})
-               list (APPEND dlib_needed_public_libraries ${X11_LIBRARIES})
-            else()
-               find_library(xlib X11)
-               # Make sure X11 is in the include path.  Note that we look for
-               # Xlocale.h rather than Xlib.h because it avoids finding a partial
-               # copy of the X11 headers on systems with anaconda installed.
-               find_path(xlib_path Xlocale.h
-                  PATHS 
-                  /Developer/SDKs/MacOSX10.4u.sdk/usr/X11R6/include
-                  /opt/local/include
-                  PATH_SUFFIXES X11
-                  )
-               if (xlib AND xlib_path)
-                  get_filename_component(x11_path ${xlib_path} PATH CACHE)
-                  list (APPEND dlib_needed_public_includes ${x11_path})
-                  list (APPEND dlib_needed_public_libraries ${xlib})
-                  set(X11_FOUND 1)
-               endif()
-            endif()
-            if (NOT X11_FOUND)
-               message(" *****************************************************************************")
-               message(" *** DLIB GUI SUPPORT DISABLED BECAUSE X11 DEVELOPMENT LIBRARIES NOT FOUND ***")
-               message(" *** Make sure XQuartz is installed if you want GUI support.               ***")
-               message(" *** You can download XQuartz from: https://www.xquartz.org/               ***")
-               message(" *****************************************************************************")
```

Comment:
Same for this message, this is super useful to people.  Otherwise they will just have no idea.

Reply:
The cmake module for X11 (FindX11.cmake) does notify you if X11 isn't found. Yes it's not giving you a message that is telling what to install but it's still there. It might even check the header I'm not sure.

My view is, times have moved on, cmake is the de facto build standard for C++ and users know how to interpret it or at least google the right stuff. Maybe in 2006 that wasn't the case but now I expect people to know what to do when they read (X11_NOTFOUND missing header...).

The main motivations for this PR are to clean things up, remove deprecated code, and reduce lines of code where possible. And using modern cmake can help with that. I consider lots of messages as noise personally. 

Maybe we need more runners that test edge cases that you're worried about. Then we can see the cmake logs and judge whether or not they are helpful enough. 

But at the end of the day, this is your library so we'll do whatever you want.

Reply:
Na times haven't moved on.  People still post constantly in various places, and email me, about not being able to compile dlib.  Even though dlib is super easy to compile. The problem is python users and especially windows python users.  If anything goes wrong they are just lost and don't read the cmake output.  So there have to be huge obvious eye catching messages in the build output telling them what to do or they email me about it.

Like to python people, even the idea that they need to have a compiler installed is not a concept they really understand. For instance, this is a hugely useful message:
```
if (WIN32 AND NOT "${CMAKE_GENERATOR}" MATCHES "Visual Studio")
   message(FATAL_ERROR "\n"
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n"
      "You must use Visual Studio to build a python extension on windows. If you "
      "are getting this error it means you have not installed Visual C++.  Note that "
      "there are many flavors of Visual Studio, like Visual Studio for C\# development. " 
      "You need to install Visual Studio for C++. \n"
      "!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n")
endif()
```

Which at some level is a ridiculous thing to have to point out to people. And yet, for the typical python user they may not have even ever heard of visual studio for C++.  Lots of them will go install visual studio but not click the option to install the C++ compiler and then be confused why "it's not working".  There are no ends of questions about that like "but I installed visual studio?".  Even with this warning people ask about it sometimes still.

So it's not about needing more test runners or anything.  It's just user education and very contextual docs.

Reply:
Like man, it's really python users.  Python users are the bane of my existence.

Reply:
Maybe you need to remove your email from the dlib website :) 
And if there is a way to auto delete github issues that don't conform to a template with reproducible code then do that.

Reply:
Ha, na.  We should just make the tooling as nice and easy to use as it can be.  It is no hardship to have messages like these and they help a huge number of people.